### ISSUE TITLE
RPC server: Implement backpressure for subscriptions

### ISSUE THREAD

**Issue by lexnv** [paritytech/substrate#11689]
## Details

RPC subscriptions produce items that are unbounded. This can lead to internal buffers growing too much,
especially when consumers are too slow.

This issue opens the discussion for implementing backpressure for the RPC subscriptions.

### JsonRpsee + Substrate

One option for implementing this would be to have a type of `BoundedStream` implemented in the JsonRPSee
which ensures that the subscription producer is bounded (ie buffered with a limit). As a side effect,
the substrate implementation part would need to change to provide backpressure.

This would change the API of the JsonRpsee with regard to subscriptions.

At its core, the subscriptions rely on their `sink` to `pipe_from_stream` data back to the consumers.

```rust
fn subscribe_justifications(&self, pending: PendingSubscription) {
  let stream = self.justification_stream.subscribe().map(
	  |x: sc_finality_grandpa::GrandpaJustification<Block>| {
		  JustificationNotification::from(x)
	  },
  );
  
  let fut = async move {
	  if let Some(mut sink) = pending.accept() {
		  sink.pipe_from_stream(stream).await;
	  }
  };
  
  self.executor.spawn("substrate-rpc-subscription", Some("rpc"), fut.boxed());
}
```

This would imply that the `pipe_from_stream` would accept a `BoundedStream` that the `justification_stream` (ie `GrandpaJustificationStream`) would need to comply.

### Substrate

Another option would be to have support for backpressure entirely in the substrate. This would ensure that the
JsonRpsee interface remains unchanged. Although the work is similar in the substrate, ie providing backpressure 
does not change for this option, the JsonRpsee interface no longer enforces backpressure. A downside of this
would be that some subscriptions implement backpressure, while others might not.

## Previous Work
- https://github.com/paritytech/substrate/issues/6096 - proposed the backpressure for all unbounded queues
- https://github.com/paritytech/polkadot-sdk/issues/47 - a static alternative to limiting the subscriptions using the resource limiting functionality

**Comment by lexnv** [comment]
cc @niklasad1 @jsdw @koute

**Comment by koute** [comment]
At least for the storage change notifications my PR here adds backpressure support to `substrate`: https://github.com/paritytech/substrate/pull/11382

With that PR merged any pending storage notifications that get consumed too slowly get automatically dropped (but in a way where the client won't actually miss any notifications, so it's different than just blindly dropping them). I haven't looked into other subscriptions we have, but I imagine we'll have to consider how to handle those on a case-by-case basis.

**Comment by koute** [comment]
So just to make sure: if the client's too slow to consume the messages then `jsonspsee` will currently just infinitely buffer them, correct?

> I'd like to know whether there is something that we can provide on the `jsonrpsee` side of things that would make it easier for people to do this stuff, though?

This might not necessarily make it easier, but when creating a subscription handler I think API-wise it'd make sense to be forced by `jsonrpsee` to either:

a) always specify a limit of how many messages `jsonrpsee` should buffer (or not buffer at all and just depend on the caller to do all of the buffering),
b) or *explicitly* have to opt into an unbound buffer.

Basically make it so that there won't be any confusion about what's bounded and what's not, and that any such decisions will have to be made explicitly and will have to be documented by the user at the call site.

**Comment by tomaka** [comment]
cc https://paritytech.github.io/json-rpc-interface-spec/dos-attacks-resilience.html for a quick 101 about back-pressure
